### PR TITLE
feat: add support for go workspaces

### LINKED ISSUE #73: Support Go 1.18 workspaces
# NX-GO : Support Go 1.18 workspaces
Go 1.18 introduces Go workspaces, a clean and nice way to manage a repository that contains multiple modules written in go. It works by adding a `go.wok` file in the root directory and add a `go.mod` file in each go module.

To adapt this solution to nx-go, we need to be able to manage versions before 1.18 (without the `go.work` files) and the next versions.

It can be also nice to add an utility to upgrade the current workspaces to th new module management system.

## App & libs creation
When creating apps and libs we have now 4 use cases. (The first match skip the followings)
- A `go.mod` file is in the root directory: the new package is created as before. 
- A `go.work` file is in the root directory: the package is created with its own `go.mod` file in its directory and it is added in the `use` list in the `go.work` file.
- The option "useWorkspace" is added in the command line the go version is checked, if it is before 1.18 an error is thrown otherwise, a `go.work` file is added in the root directory.
- Same behavior as before 1.18

### PR DESCRIPTION
**Pull request by Anthony-Jhoiro**
closes #73

### CODE REVIEW

File: packages/nx-go/src/utils/update-go-work.helper.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+export function updateGoWork(tree: Tree, options: NormalizedSchema) {
+  if (options.useGoWork) {
+    if (!tree.exists(GO_WORK_FILE)) {
+      if (!canUseWokspaces()) {
+        throw new Error('Your version of go does not support workspaces')
+      }
+
+      const [major, minor] = getGoVersion()
```
Comment (lewsmith) [inline packages/nx-go/src/utils/update-go-work.helper.ts:34]:
```suggestion
      const [major, minor] = getGoVersion().split('.')
```

**Comment by Afourcat** [comment]
Hello,
Thank you for your work.

This is a feature that we are waiting for, @Anthony-Jhoiro, do you need some help?
If not, does someone know when this could be live?

Thank you again,

Alexandre,

**Comment by Anthony-Jhoiro** [comment]
Hello @Afourcat ! 

Sorry, I've been quite busy lately, but I will get back to work on this shortly. I'll try to mark this draft as ready for review by the end of next week :crossed_fingers:. For the moment, the generation seems to be working well. I still need to work on the dependency graph and check that the executors are working fine.

File: packages/nx-go/src/utils/update-go-work.helper.ts (hunk 1 above)
Comment (Anthony-Jhoiro) [reply packages/nx-go/src/utils/update-go-work.helper.ts]: Thanks, I forgot to add it!

**Comment by Afourcat** [comment]
Hello,
Thank you @Anthony-Jhoiro 🦾
Is this ready for upstream? if no, is there something I can do?

**Comment by beeman** [comment]
A heads-up from my side... I've been traveling quite a bit over the last months so had little time for my OSS projects, but I'm back home and ready to work on this project in my weekend! 🎉

**Comment by Anthony-Jhoiro** [comment]
Hello @Afourcat It seems ready for me! You can review it if you want :wink:

**Comment by Afourcat** [comment]
> Hello @Afourcat It seems ready for me! You can review it if you want :wink: 

Hello, I'm not a maintainer of the project, just an interested user 😅.
I'm fully relying on you guys 💪.

File: packages/nx-go/src/generators/library/generator.ts [hunk 2]
```diff
@@ -22,5 +22,11 @@ export default async function (tree: Tree, options: LibraryGeneratorSchema) {
     tags: normalizedOptions.parsedTags,
   })
   addFiles(tree, join(__dirname, 'files'), normalizedOptions)
+
+  if (normalizedOptions.useGoWork) {
+    createGoMod(tree, normalizedOptions)
```
Comment (beeman) [inline packages/nx-go/src/generators/library/generator.ts:27]: Do we need to call `createGoMod` when creating a library?

**Comment by beeman** [comment]
@Anthony-Jhoiro all looking very good here, just one question about the `createGoMod` methoad.

I've just pushed `@nx-go/nx-go@2.7.0-rc.0` that this PR, so feel free to give it a spin.

File: packages/nx-go/src/utils/run-go-command.helper.ts [hunk 3]
```diff
@@ -16,10 +16,10 @@ export function runGoCommand(
 
   try {
     console.log(`Executing command: ${execute}`)
-    execSync(execute, { cwd, stdio: [0, 1, 2] })
```
Comment (beeman) [inline packages/nx-go/src/utils/run-go-command.helper.ts:19]: FYI it turns out that this breaks the usage of any of the go commands except `version` as the output no longer gets piped to the terminal. 

I'm patching this in #76

File: packages/nx-go/src/generators/library/generator.ts (hunk 2 above)
Comment (beeman) [reply packages/nx-go/src/generators/library/generator.ts:27]: I think I already understand why this is, we need a `go.mod` in all the separate projects of the workspace, is that correct?

**Comment by beeman** [comment]
Landed in #76 - thanks a lot for your work on this @Anthony-Jhoiro 🥳

**Pull request closed without merging**